Makefile: fix targets `merge-kernel-configs` and `full-config`

sky1122 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #174

**Description of changes:**
Refactored Makefile targets merge-kernel-configs and full-config to follow standard Make idioms. Cleaned up latest-kernel-full-config.sh script, corrected environment variables for cross-compilation support and added check for running only in SDK container.

I keep creating the temp directory under the target so that I can clean the temp directory after exit. 


**Testing done:**
Generating the config-full-bottlerocket-aarch64 and config-full-bottlerocket-x86_64 for 6.1 kernel and diff with the one in repo


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

sky1122 committed:
makefile: refactor full-config target to use native make constructs

The full-config target was rewritten to use Make's built-in functions
instead of complex shell scripting.

Signed-off-by: Jingwei Wang <jweiw@amazon.com>

sky1122 commented:
force pushed to rebase

arnaldo2792 commented on Makefile:
```diff
@@ -23,65 +57,40 @@ endif
-	fi; \
-	sdk_image=$$(source $(KERNEL_CONFIG_SCRIPT) && resolve_bottlerocket_sdk); \
+	@test $${RPM_FILE?not set!}
+	@echo "Checking kernel configuration dependencies..."
+	$(eval MISSING_TOOLS := $(shell for tool in rpm2cpio cpio tar docker tq; do command -v $$tool >/dev/null 2>&1 || echo $$tool; done))
+	$(if $(MISSING_TOOLS),$(error DEPENDENCY ERROR: Missing tools: $(MISSING_TOOLS). Please install them in your PATH.))
```
This could be a target, which is a dependency of this target.

arnaldo2792 commented on Makefile:
```diff
@@ -23,65 +57,40 @@ endif
-	sdk_image=$$(source $(KERNEL_CONFIG_SCRIPT) && resolve_bottlerocket_sdk); \
+	@test $${RPM_FILE?not set!}
+	@echo "Checking kernel configuration dependencies..."
+	$(eval MISSING_TOOLS := $(shell for tool in rpm2cpio cpio tar docker tq; do command -v $$tool >/dev/null 2>&1 || echo $$tool; done))
```
You are doing your validation on at least `tq`, after you have already attempted to use it in the first section of your Makefile.

arnaldo2792 commented on Makefile:
```diff
@@ -5,6 +5,40 @@ TWOLITER := $(TWOLITER_DIR)/twoliter
+SDK_REGISTRY_OVERRIDE := $(shell tq -r ".bottlerocket.bottlerocket-sdk.registry" -f Twoliter.override 2>/dev/null)
+SDK_NAME_OVERRIDE := $(shell tq -r ".bottlerocket.bottlerocket-sdk.name" -f Twoliter.override 2>/dev/null)
+
+# Use override values if available, otherwise use defaults
+SDK_REGISTRY := $(if $(SDK_REGISTRY_OVERRIDE),$(SDK_REGISTRY_OVERRIDE),$(SDK_REGISTRY_DEFAULT))
+SDK_NAME := $(if $(SDK_NAME_OVERRIDE),$(SDK_NAME_OVERRIDE),$(SDK_NAME_DEFAULT))
```
Don't you have to set this after `SDK_REGISTRY_DEFAULT`?

sky1122 replied:
putting in after checking [Twoliter.override](https://github.com/bottlerocket-os/bottlerocket-kernel-kit/pull/258/files/dcb742307edb88a777aa4bdb0510a287dc14e3ee#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R15-R16) to make sure if there is a Twoliter.override so that we get the sdk value from that. Similar with the [original script](https://github.com/bottlerocket-os/bottlerocket-kernel-kit/blob/develop/tools/latest-kernel-full-config.sh#L54C1-L54C25)

sky1122 replied:
I have some repeat code probably from rebase. delete them....

arnaldo2792 commented on Makefile:
```diff
@@ -23,65 +57,40 @@ endif
-	validate_param "KVER" "$(KVER)" "KVER parameter is required"; \
-	echo "All parameter checks passed. Merging kernel configs for version $(KVER)..."
-
+	@test $${SDK_IMAGE?not set!}
```
Shouldn't this be the `RESOLVED_SDK_IMAGE`?

sky1122 replied:
My purpose is using `full-config` to pass this `SDK_IMAGE` to `merge-kernel-configs` . That's why I didn't directly use `RESOLVED_SDK_IMAGE `
```
$(MAKE) merge-kernel-configs RPM_FILE="$${rpm_file}" SDK_IMAGE="$(RESOLVED_SDK_IMAGE)" KVER="$${kver}"
```

Let me check how to make this better by following the next comment

arnaldo2792 commented on Makefile:
```diff
@@ -23,65 +57,40 @@ endif
-	echo "Using SDK image: $${sdk_image}"; \
-	echo "Using RPM file: $${rpm_file}"; \
-	$(MAKE) merge-kernel-configs RPM_FILE="$${rpm_file}" SDK_IMAGE="$${sdk_image}" KVER="$${kver}"
+	$(MAKE) merge-kernel-configs RPM_FILE="$${rpm_file}" SDK_IMAGE="$(RESOLVED_SDK_IMAGE)" KVER="$${kver}"
```
This is weird, you are calling a make target from within a make target. You should do this instead:

```Makefile
# Call merge-kernel-configs before full-config is called
full-config: merge-kernel-configs
    # Do the full-config stuff here
```

sky1122 replied:
I was thinking this can be separated logic for the `full-config`. But after revision, I think I can combine these two in one function based on the fact `full-config` was doing dependency check and getting the parameters that `merge-kernel-configs` needed.

sky1122 commented:
force push to address feedback. Main change
- Moving resolve sdk image logic back to `latest-kernel-full-config.sh` since IMO it looks more clear for the Makefile
- move the dependency check to a target
- merge `full-config` and `merge-kernel-configs` to one target.

sky1122 committed:
tools: clean up latest-kernel-full-config.sh

Removing functions that were moved to the Makefile.
Environment validation updated to check for Docker
mount points.

Signed-off-by: Jingwei Wang <jweiw@amazon.com>

sky1122 commented:
force pushed to fix the commit message

sky1122 commented:
force push to rebase

ginglis13 commented on Makefile:
```diff
@@ -22,66 +22,38 @@ endif
-
+check-dependencies:
+	@echo "Checking kernel configuration dependencies..."
+	$(eval MISSING_TOOLS := $(shell for tool in rpm2cpio cpio tar docker tq; do command -v $$tool >/dev/null 2>&1 || echo $$tool; done))
```
now that we've containerized things, aren't we getting `rpm2cpio` and `cpio` from the SDK image?

sky1122 replied:
yes. `tq` and `docker` need to have. will change that

sky1122 commented:
force pushed to change the dependency check context.
force pushed to make the commits have correct content.

arnaldo2792 commented on Makefile:
```diff
@@ -22,66 +22,38 @@ endif
-	validate_param "KVER" "$(KVER)" "KVER parameter is required"; \
-	echo "All parameter checks passed. Merging kernel configs for version $(KVER)..."
-
+check-dependencies:
```
I talked with @sky1122 offline. The idea is to make this happen:

```Makefile
full-config:
    docker run <whatever> full-config
```

So that all the setup happens in the container, and all the evaluations/validations occur in the script that is executed in the container.

sky1122 commented:
Here is the reason why I still keep these shell command in the Makefile:
- need `RPM_PATH` to get the kernel-xx-rpm package
- SDK_IMAGE: need to get to pass to docker as the image name
- KVER: need to auto resolve from the rpm package. Only run `full-config` when a related kernel rpm provided. not always running all
- KERNEL_PACKAGE_DIR: need to move the final full-config outside the directory

arnaldo2792 commented on Makefile:
```diff
@@ -22,66 +22,38 @@ endif
+	$(eval export RPM_PATH := $(shell realpath "$(RPM_FILE)"))
+	$(eval export SDK_IMAGE := $(shell cd $(TOP) && source $(KERNEL_CONFIG_SCRIPT) && resolve_bottlerocket_sdk))
+	$(eval export KVER := $(shell rpm --query --nosignature --queryformat '%{VERSION}' "$(RPM_FILE)" | sed 's/\.[^.]*$$//'))
+	$(eval export KERNEL_PACKAGE_DIR := $(TOP)packages/kernel-$(KVER))
```
You can do all of this within the container, if you mount `$(TOP)/packages/`

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +97,91 @@ validate_kernel_configs() {
+
+# Create temporary working directory
+tmpdir=$(mktemp -d)
+trap 'rm -rf ${tmpdir}' EXIT
```
This line should be on top of the file, instead of here, and should call a cleanup function rather than the once command. But since you are running this in a container, why do you care about the clean up? The filesystem of the container is ephemeral so you don't really have to clean up anything.

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +97,91 @@ validate_kernel_configs() {
+popd || bail "Could not move around - 'popd' back to /work failed. Lets stop before we break anything further."
+
+for arch in "x86_64" "aarch64"; do
+
```
The extra space isn't really needed.

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +97,91 @@ validate_kernel_configs() {
+        script_args=("../config-${arch}" "${microcode_cfg}" "${br_cfg}" "${br_cfg_arch}")
+    fi
+
+    ARCH=${karch} CROSS_COMPILE=/usr/bin/${arch}-bottlerocket-linux-gnu- KCONFIG_CONFIG=bottlerocket_${arch}_defconfig ./scripts/kconfig/merge_config.sh "${script_args[@]}"
```
You could break down this line to make the script more readable (try to keep your line length under 85 characters)

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +97,91 @@ validate_kernel_configs() {
 }
+
+# Only check container environment if we're running the main script (not just sourcing functions)
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
```
I think you can skip the check to validate if the script is running inside a container. Instead you can perform all the validations that you require for the script to work properly, like you did underneath.

sky1122 committed:
tools: add docker-run script for SDK container execution

The script executes build scripts within the Bottlerocket SDK container
by reading configuration from Twoliter.lock and optional overrides from
Twoliter.override. It mounts the current directory and preserves user
permissions to ensure proper file ownership of generated artifacts.

Signed-off-by: Jingwei Wang <jweiw@amazon.com>

sky1122 commented:
forced push to fix the commit history and remove mount rpm file into SDK container

sky1122 commented:
force pushed to allow when only one kernel has rpm package

arnaldo2792 commented on tools/docker-run.sh:
```diff
@@ -0,0 +1,32 @@
+docker run --rm \
+    -v "$(pwd):/bottlerocket-kernel-kit" \
+    --user "$(id -u):$(id -g)" \
+    --name "kernel-full-config" \
```
You don't have to give this a name, if anything, it shouldn't be always `kernel-full-config` as the purpose of the script is to be used to run any script within a docker container

arnaldo2792 commented on tools/docker-run.sh:
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
```
nit: in your commit message, you could provide a bit more details on this new script

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +160,56 @@ validate_kernel_configs() {
+
+# Check which kernels have RPMs available and process only those
+kernels_to_process=()
+for kernel_pkg in "kernel-6.1" "kernel-6.12"; do
```
What if instead of hardcoding the list of kernels, you just generate the configs for all the kernels that you found in the packages directory? That way, we don't have to keep updating the script as we add more kernels.

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -10,82 +10,31 @@ bail() {
 }
 
-# Function to display usage information
-usage() {
```
IMO, you still need the `usage` function, probably a bit simplified, at least to explain people that mistakenly ran the script `latest-kernel-full-config.sh` without the `Makefile` target.

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -10,82 +10,31 @@ bail() {
-inner_full_config() {
-    pushd /work || bail "Unable to enter /work (bind mount to tmp dir)"
+    local majorminor=${version%.*} # Trim after last '.', e.g. 6.1.132 -> 6.1
+    local KERNEL_PACKAGE_DIR="/bottlerocket-kernel-kit/packages/kernel-${majorminor}"
```
capitalized variables are meant to be used for constants, not local variables.

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -10,82 +10,31 @@ bail() {
-    >&2 usage
-    bail "$1"
-}
+MICROCODE_DIR="/bottlerocket-kernel-kit/packages/microcode"
```
Constants go on the top of the file

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +160,56 @@ validate_kernel_configs() {
 }
+
+# Set paths
+KERNEL_KIT_DIR="/bottlerocket-kernel-kit"
```
Same as elsewhere, constants on the top

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +184,79 @@ validate_kernel_configs() {
+    echo "   Expected mount point ${KERNEL_KIT_DIR} not found."
+    echo ""
+    echo "   Please use: make full-config"
+    echo ""
```
You should just use `usage` here.

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +184,79 @@ validate_kernel_configs() {
+    bail "No kernel RPMs found in any package directory"
+fi
+
+# Process available kernels
```
Comments should explain why you did what you did, not what's happening. The code is explicit enough to understand what's happening. Same for all other comments.

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -2,90 +2,66 @@
-    exit 1
-}
+MICROCODE_DIR="/bottlerocket-kernel-kit/packages/microcode"
+KERNEL_KIT_DIR="/bottlerocket-kernel-kit"
```
nit:

```suggestion
KERNEL_KIT_DIR="/bottlerocket-kernel-kit"
MICROCODE_DIR="${KERNEL_KIT_DIR}/packages/microcode"
```

arnaldo2792 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -2,90 +2,66 @@
-# Function to perform the kernel configuration merging inside Docker container
-inner_full_config() {
-    pushd /work || bail "Unable to enter /work (bind mount to tmp dir)"
+    local kernel_package_dir="/bottlerocket-kernel-kit/packages/kernel-${majorminor}"
```
You could use `${KERNEL_KIT_DIR}` here

ginglis13 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -2,90 +2,66 @@
-EOF
-}
+The script will:
+1. Discover all kernel-* packages with available RPMs
```
What is the behavior when there are multiple RPMs in a kernel package subdir? This PR would force me to have some better hygiene, but currently my kernel-6.12 dir has many RPMs across a few different builds of the kernel-kit:

```
╭─fedora in ~/kernel-kit/packages/kernel-6.12  ‹develop*›
╰─➤  ls *.rpm
aws-neuronx-dkms-2.20.28.0.noarch.rpm  kernel6.12-6.12.25-32.101.amzn2023.src.rpm  kernel6.12-6.12.37-61.105.amzn2023.src.rpm  kernel6.12-6.12.40-64.114.amzn2023.src.rpm
aws-neuronx-dkms-2.21.37.0.noarch.rpm  kernel6.12-6.12.29-33.102.amzn2023.src.rpm  kernel6.12-6.12.40-63.114.amzn2023.src.rpm
```

ginglis13 commented on tools/latest-kernel-full-config.sh:
```diff
@@ -213,3 +184,71 @@ validate_kernel_configs() {
+for kernel_dir in "${KERNEL_KIT_DIR}/packages"/kernel-*; do
+    if [[ -d "${kernel_dir}" ]]; then
+        kernel_pkg=$(basename "${kernel_dir}")
+        rpm_file=$(find "${kernel_dir}" -name "kernel*.src.rpm" | head -1)
```
This should get the latest RPM if you use `tail`:


```suggestion
        rpm_file=$(find "${kernel_dir}" -name "kernel*.src.rpm" | tail -1)
```
